api/types/authentication_test.go

```diff
@@ -64,3 +68,144 @@ func TestEncodeDecodeRequireMFAType(t *testing.T) {
 		})
 	}
 }
+
+func TestAuthPreference_CheckAndSetDefaults(t *testing.T) {
+	for _, tt := range []struct {
+		name           string
+		spec           AuthPreferenceSpecV2
+		assertErr      require.ErrorAssertionFunc
+		assertAuthPref func(t *testing.T, authPref AuthPreference)
+	}{
+		{
+			name: "OK default to OTP",
+			spec: AuthPreferenceSpecV2{},
+			assertAuthPref: func(t *testing.T, authPref AuthPreference) {
+				require.Equal(t, []SecondFactorType{SecondFactorType_SECOND_FACTOR_TYPE_OTP}, authPref.GetSecondFactors())
+			},
+		},
+		{
+			name: "OK OTP",
+			spec: AuthPreferenceSpecV2{
+				SecondFactors: []SecondFactorType{
+					SecondFactorType_SECOND_FACTOR_TYPE_OTP,
+				},
+			},
+			assertAuthPref: func(t *testing.T, authPref AuthPreference) {
+				require.False(t, authPref.GetAllowPasswordless())
+				require.False(t, authPref.GetAllowHeadless())
+			},
+		},
+		{
+			name: "OK WebAuthn",
+			spec: AuthPreferenceSpecV2{
+				SecondFactors: []SecondFactorType{
+					SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN,
+				},
+				Webauthn: &Webauthn{
+					RPID: "localhost",
+				},
+			},
+			assertAuthPref: func(t *testing.T, authPref AuthPreference) {
+				require.True(t, authPref.GetAllowPasswordless())
+				require.True(t, authPref.GetAllowHeadless())
+			},
+		},
+		{
+			name: "OK SSO",
+			spec: AuthPreferenceSpecV2{
+				SecondFactors: []SecondFactorType{
+					SecondFactorType_SECOND_FACTOR_TYPE_SSO,
+				},
+				AllowLocalAuth: NewBoolOption(false),
+			},
+			assertAuthPref: func(t *testing.T, authPref AuthPreference) {
+				require.False(t, authPref.GetAllowPasswordless())
+				require.False(t, authPref.GetAllowHeadless())
+				require.False(t, authPref.GetAllowLocalAuth())
+			},
+		},
+		{
+			name: "OK U2F config provided",
+			spec: AuthPreferenceSpecV2{
+				SecondFactors: []SecondFactorType{
+					SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN,
+				},
+				U2F: &U2F{
+					AppID: "https://localhost",
+				},
+			},
+		},
+		{
+			name: "NOK SecondFactor and SecondFactors both set",
+			spec: AuthPreferenceSpecV2{
+				SecondFactor: constants.SecondFactorWebauthn,
+				SecondFactors: []SecondFactorType{
+					SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN,
+				},
+			},
+			assertErr: func(t require.TestingT, err error, vals ...interface{}) {
+				require.ErrorAs(t, err, &trace.BadParameterError{})
```

Comment:
That's a bit strange, I would expected we needed to pass `**trace.BadParameterError` to errors.As, since it's `*trace.BadParameter` that implements `error`.